src/Compilers/Core/Portable/Diagnostic/Diagnostic.cs

```diff
@@ -139,15 +139,56 @@ public abstract partial class Diagnostic : IEquatable<Diagnostic>, IFormattable
                 throw new ArgumentNullException(nameof(descriptor));
             }
 
-            var warningLevel = GetDefaultWarningLevel(effectiveSeverity);
+            return Create(
+                descriptor,
+                location,
+                effectiveSeverity.ToReportDiagnostic(),
+                additionalLocations,
+                properties,
+                messageArgs);
+        }
+
+        /// <summary>
+        /// Creates a <see cref="Diagnostic"/> instance.
+        /// </summary>
+        /// <param name="descriptor">A <see cref="DiagnosticDescriptor"/> describing the diagnostic.</param>
+        /// <param name="location">An optional primary location of the diagnostic. If null, <see cref="Location"/> will return <see cref="Location.None"/>.</param>
+        /// <param name="effectiveSeverity">Effective severity of the diagnostic.</param>
+        /// <param name="additionalLocations">
+        /// An optional set of additional locations related to the diagnostic.
+        /// Typically, these are locations of other items referenced in the message.
+        /// If null, <see cref="AdditionalLocations"/> will return an empty list.
+        /// </param>
+        /// <param name="properties">
+        /// An optional set of name-value pairs by means of which the analyzer that creates the diagnostic
+        /// can convey more detailed information to the fixer. If null, <see cref="Properties"/> will return
+        /// <see cref="ImmutableDictionary{TKey, TValue}.Empty"/>.
+        /// </param>
+        /// <param name="messageArgs">Arguments to the message of the diagnostic.</param>
+        /// <returns>The <see cref="Diagnostic"/> instance.</returns>
+        public static Diagnostic Create(
+            DiagnosticDescriptor descriptor,
+            Location location,
+            ReportDiagnostic effectiveSeverity,
```

Comment:
>ReportDiagnostic effectiveSeverity [](start = 12, length = 34)

I am not sure about the current semantics of the added API. `ReportDiagnostic` is meant only to be used for rule ID configuration purposes, and is never tied to a specific diagnostic instance. For example, there is no real semantic meaning of creating a diagnostic instance with `ReportDiagnostic.Suppressed`, which indicates this rule ID has been disabled.

Reply:
That is not the meaning of `ReportDiagnostic.Suppress` - it is incorrectly named as `Suppress` instead of `Disable`, which is the root of the confusion. We never create diagnostic instances for rule IDs that have been disabled by user configuration. For specific diagnostic instances that have been suppressed with source suppressions, we create a diagnostic instance with `IsSuppressed` flag set to true. This subtle difference is used by the IDE error list feature that enables you to view and audit source suppressions (see `Suppression State` column in error list).

Reply:
See http://source.roslyn.io/#Microsoft.CodeAnalysis/Diagnostic/Diagnostic.cs,327, the doc comment has the precise definition for the semantics of this flag, which should have probably been named as `IsSuppressedInSource` or something.

Reply:
I'm not particularly happy with the alternatives here, especially given we both want to see wider improvements (e.g. discussion in https://github.com/dotnet/roslyn/pull/27220#discussion_r192830806). The only clear drawback for me on the current approach is it adds a public API where we aren't sure of the long-term vision. If there is a redeeming factor, it would likely be considering the new overload as a natural simplification of another public overload:

https://github.com/dotnet/roslyn/blob/969249c9a2bf6ab6ded98d9e7702e76d5ebd6bf1/src/Compilers/Core/Portable/PublicAPI.Shipped.txt#L2673

I'd like to have a brainstorming discussion with you today and see if we can find any other alternatives to consider.

Reply:
Yes, let us sync offline later today. We cannot generate a diagnostic instance for `ReportDiagnostic.Suppress`, as it breaks a bunch of existing semantics (`IsSuppressed` no longer represents source suppressions, our diagnostic system now starts producing diagnostic instances for disabled rules) and error list features (use of `Suppression State` column to view source suppressed diagnostics and use the error list context menus to toggle between removing source suppressions and adding them back).

Reply:
Yeah I went down this path. It doesn't work.